### PR TITLE
Minimise object allocation in MappingMetadata parsing

### PR DESCRIPTION
**Pull request by romseygeek**
Currently, when building a `MappingMetadata` object from x-content, we build
a realised `Map<String, Object>` by calling `parser.map()` and then re-serialize
this in the `MappingMetadata` constructor.  This can create a lot of small objects
that are immediately thrown away, particularly for large mappings, which seems
wasteful.

This commit changes `IndexMetadata` x-content parsing code to instead pass
the mappings map directly as a `BytesReference`.  We still need to do some 
parsing but we avoid much of the object allocation.  It also tightens up the contract
for what `MappingMetadata` will accept as its source, which must always be in
json format and have a containing type field.

### CODE REVIEW

**Comment by romseygeek** [comment]
@elasticmachine update branch

**Comment by romseygeek** [comment]
@elasticmachine run elasticsearch-ci/2

File: server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java [hunk 1]
```diff
@@ -58,53 +61,67 @@ public MappingMetadata(DocumentMapper docMapper) {
...
     }
 
-    @SuppressWarnings("unchecked")
-    public MappingMetadata(CompressedXContent mapping) {
-        this.source = mapping;
-        Map<String, Object> mappingMap = XContentHelper.convertToMap(mapping.compressedReference(), true).v2();
-        if (mappingMap.size() != 1) {
-            throw new IllegalStateException("Can't derive type from mapping, no root type: " + mapping.string());
+    // xcontent is always of type JSON and wrapped in a type object
+    public MappingMetadata(BytesReference mapping) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MappingMetadata.java:65]: What if were to keep the `CompressedXContent` constructor instead here? I would feel more comfortable about that, as we currently implicitly assume that the bytes passed here are always decompressed?

Note that we are anyway decompressing the array again in the `new CompressedXContent()` call in line 83 as `CompressedXContent` is computing the `crc32`.

**Review by ywelsch — CHANGES_REQUESTED** [review]

**Comment by ywelsch** [comment]
@romseygeek are you still interested in pursuing this PR?

**Comment by romseygeek** [comment]
@ywelsch I am, but I don't think I'm going to have time to look at it properly any time soon

**Comment by ywelsch** [comment]
ok, let's close the PR for the time being then.

**Pull request closed without merging**